### DISCUSSION TITLE
VS 2022 Arm64 Support

### DISCUSSION THREAD

**Discussion by CoreyKaylor** [ormsolutions/NORMA#45] [Ideas]
Are there plans to release a version that's compatible for 2022 on the marketplace? I noticed it's only available for the 2019 version. BTW @mcurland, I hope you're doing well. It's been a while. :smile:

**Comment by mcurland** [comment]
Hi Corey, good to hear from you.

I'm not sure why to you're only seeing the 2019 version. There are admin (all user) and per user installations for VS2017/VS219/VS2022.

If you go into the online marketplace from the VS extension manager you'll match the right version. They all look similar from the Marketplace website--just adjust the URL.

https://marketplace.visualstudio.com/items?itemName=ORMSolutions.NORMA2022

https://marketplace.visualstudio.com/items?itemName=ORMSolutions.NORMA2022PerUser

**Comment by CoreyKaylor** [comment]
Please see what shows up when I search for it from the VS2022 Manage Extensions dialog. For the 2019 experience, I was able to find it by searching similar to the first example and install the extension.
<img width="1540" alt="Screenshot 2023-03-18 at 7 11 18 AM" src="https://user-images.githubusercontent.com/121507/226101825-4be3ed09-1e61-4efd-be88-aabc00b27ff5.png">

<img width="1578" alt="Screenshot 2023-03-18 at 7 11 54 AM" src="https://user-images.githubusercontent.com/121507/226101829-d5e41f1f-cdac-49f6-a3f7-5f107075fd23.png">

Even if I manually download the extension from the links you've provided. Here's the message and logs for what happens when trying to install.

<img width="538" alt="Screenshot 2023-03-18 at 7 14 38 AM" src="https://user-images.githubusercontent.com/121507/226101950-b2ba4570-5253-4180-9903-dfe67daad952.png">

3/18/2023 7:14:12 AM - Microsoft VSIX Installer
3/18/2023 7:14:12 AM - -------------------------------------------
3/18/2023 7:14:12 AM - vsixinstaller.exe version:
3/18/2023 7:14:12 AM - 17.5.4065+gbda5f544dc
3/18/2023 7:14:12 AM - -------------------------------------------
3/18/2023 7:14:12 AM - Command line parameters:
3/18/2023 7:14:12 AM - C:\Program Files (x86)\Microsoft Visual Studio\Installer\resources\app\ServiceHub\Services\Microsoft.VisualStudio.Setup.Service\VSIXInstaller.exe,\\Mac\Home\Downloads\NORMA.VS2022.PerUser.vsix
3/18/2023 7:14:12 AM - -------------------------------------------
3/18/2023 7:14:12 AM - Microsoft VSIX Installer
3/18/2023 7:14:12 AM - -------------------------------------------
3/18/2023 7:14:15 AM - Initializing Install...
3/18/2023 7:14:17 AM - Extension Details...
3/18/2023 7:14:17 AM - 	Identifier         : 3c58bedd-dfd4-4910-80f8-5bc564257f05
3/18/2023 7:14:17 AM - 	Name               : Natural ORM Architect (Per User)
3/18/2023 7:14:17 AM - 	Author             : ORM Solutions
3/18/2023 7:14:17 AM - 	Version            : 1.0.3048.0
3/18/2023 7:14:17 AM - 	Description        : Object-Role Modeling Designer
3/18/2023 7:14:17 AM - 	Locale             : en-US
3/18/2023 7:14:17 AM - 	MoreInfoURL        : https://github.com/ormsolutions/NORMA
3/18/2023 7:14:17 AM - 	InstalledByMSI     : False
3/18/2023 7:14:17 AM - 	SupportedFrameworkVersionRange : [4.7.2,)
3/18/2023 7:14:17 AM - 	SignatureState     : ValidSignature
3/18/2023 7:14:17 AM - 	SignedBy           : ORM Solutions, LLC
3/18/2023 7:14:17 AM - 	Certificate Info   : 
3/18/2023 7:14:17 AM - 		-------------------------------------------------------
3/18/2023 7:14:17 AM - 		[Subject]       : CN="ORM Solutions, LLC", O="ORM Solutions, LLC", L=Draper, S=Utah, C=US
3/18/2023 7:14:17 AM - 		[Issuer]        : CN=Sectigo RSA Code Signing CA, O=Sectigo Limited, L=Salford, S=Greater Manchester, C=GB
3/18/2023 7:14:17 AM - 		[Serial Number] : 009D1920234ACC832228820B9E2CC42339
3/18/2023 7:14:17 AM - 		[Not Before]    : 4/28/2021 8:00:00 PM
3/18/2023 7:14:17 AM - 		[Not After]     : 4/28/2024 7:59:59 PM
3/18/2023 7:14:17 AM - 		[Thumbprint]    : 28813429A51B35C21C8FA6C5A82CFCF5B758B12E
3/18/2023 7:14:17 AM - 	Supported Products : 
3/18/2023 7:14:17 AM - 		Microsoft.VisualStudio.Community
3/18/2023 7:14:17 AM - 			Version : [17.0,18.0)
3/18/2023 7:14:17 AM - 			ProductArchitecture : amd64
3/18/2023 7:14:17 AM - 	References         : 
3/18/2023 7:14:17 AM - 		-------------------------------------------------------
3/18/2023 7:14:17 AM - 		Identifier   : 251928d1-61d1-4970-be0e-758788d0fd10
3/18/2023 7:14:17 AM - 		Name         : PLiX Code Generator
3/18/2023 7:14:17 AM - 		Version      : [1.0.3017.8,2.0.0.0)
3/18/2023 7:14:17 AM - 		MoreInfoURL  : 
3/18/2023 7:14:17 AM - 		Nested       : Yes
3/18/2023 7:14:17 AM - 	Prerequisites      : 
3/18/2023 7:14:17 AM - 		-------------------------------------------------------
3/18/2023 7:14:17 AM - 		Identifier   : Microsoft.VisualStudio.Component.CoreEditor
3/18/2023 7:14:17 AM - 		Name         : Visual Studio core editor
3/18/2023 7:14:17 AM - 		Version      : [17.0,18.0)
3/18/2023 7:14:17 AM - Signature Details...
3/18/2023 7:14:17 AM - 	Extension is signed with a valid signature.
3/18/2023 7:14:17 AM - Searching for applicable products...
3/18/2023 7:14:17 AM - Found installed product - Visual Studio Professional 2022
3/18/2023 7:14:18 AM - Found installed product - Visual Studio Professional 2019
3/18/2023 7:14:18 AM - Found installed product - Global Location
3/18/2023 7:14:18 AM - VSIXInstaller.NoApplicableSKUsException: This extension is not installable on any currently installed products.
   at VSIXInstaller.ExtensionService.GetInstallableDataImpl(IInstallableExtension extension, String extensionPackParentName, Boolean isRepairSupported, IStateData stateData, IEnumerable`1& skuData)
   at VSIXInstaller.ExtensionService.GetInstallableData(String vsixPath, String extensionPackParentName, Boolean isRepairSupported, IStateData stateData, IEnumerable`1& skuData)
   at VSIXInstaller.ExtensionPackService.IsExtensionPack(IStateData stateData, Boolean isRepairSupported)
   at VSIXInstaller.ExtensionPackService.ExpandExtensionPackToInstall(IStateData stateData, Boolean isRepairSupported)
   at VSIXInstaller.App.Initialize(Boolean isRepairSupported)
   at VSIXInstaller.App.Initialize()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()

**Reply by NielsHoogeveen** [reply]
I just installed the update of NORMA and didn't run into any issues.

**Reply by CoreyKaylor** [reply]
I'm wondering if it's because VS 2022 seems to more directly support Arm64, where as 2019 doesn't seem to have and now there is a possible precondition not being met for the installer.

<img width="952" alt="Screenshot 2023-03-18 at 8 06 30 AM" src="https://user-images.githubusercontent.com/121507/226104991-a88a7d64-f861-4505-967f-c4df3ac19ced.png">

<img width="945" alt="Screenshot 2023-03-18 at 8 05 43 AM" src="https://user-images.githubusercontent.com/121507/226104996-e2b036a4-1780-4ecd-8553-1affec95123c.png">

**Reply by CoreyKaylor** [reply]
Seems plausible:
https://devblogs.microsoft.com/visualstudio/now-introducing-arm64-support-for-vs-extensions/

**Reply by mcurland** [reply]
Hi Corey,

Yes, arm64 would do it. I haven't come close to tackling that, or frankly even noted its existence until now.

I know they encourage extension developers to put multiple versions in one package. That has just never been feasible with NORMA as there were too many version-specific dependencies to make the tool run. One of these is the GraphObject assembly that powers the DSL design surface rendering. This is a CPU-targeted assembly (not MSIL), so there needs to be an arm64 version.

Can you do me a favor before I figure out how to build and test this without acquiring an arm64 machine:
1) Under Tools/Get Tools and Features, change your installation options to include 'Visual Studio extension development'. Also, make sure 'Modeling SDK' is checked under the optional features on the right.
2) After running the installation, make sure that a few assemblies exist. In the the Common7\IDE\PrivateAssemblies directory in your VS install (you'll land in the installation root if you open a developer command prompt, some thing like C:\Program Files\Microsoft Visual Studio\2022\Community or C:\Program Files\Microsoft Visual Studio\2022\Professional).
a) dir *modeling.sdk*
b) dir *virtualtree*

You should see:
Microsoft.VisualStudio.Modeling.Sdk.Diagrams.dll
Microsoft.VisualStudio.Modeling.Sdk.Diagrams.GraphObject.dll
Microsoft.VisualStudio.Modeling.Sdk.dll
Microsoft.VisualStudio.Modeling.Sdk.Shell.dll
and
Microsoft.VisualStudio.VirtualTreeGrid.dll

If you see these assemblies then I can most likely make NORMA work, but it would be an arm64-specific version.

Just to verify that GraphObject is still CPU specific I ran the following from PowerShell in the ...\PrivateAssemblies directory:
[reflection.assemblyname]::GetAssemblyName("${pwd}\Microsoft.VisualStudio.Modeling.SDK.Diagrams.GraphObject.dll") | fl
Sure enough, the ProcessorArchitecture is Amd64. All of the other assemblies I need are MSIL. Could you verify what yours says (if the assembly is there)?

If the SDK is in place maybe we can take a few minutes and I can get you a dev build of NORMA locally? That would give me an idea on the needed changes as well running on the native system.

**Reply by CoreyKaylor** [reply]
All of them are there except for `Microsoft.VisualStudio.Modeling.Sdk.Diagrams.GraphObject.dll`.

I'm assuming we're blocked at this point, correct?

I've posted the issue here. https://developercommunity.visualstudio.com/t/MicrosoftVisualStudioModelingSdkDiag/10314287

**Reply by mcurland** [reply]
Let's give them a chance to respond before using too much more time. Do you have an amd64 box that you can run on?

Honestly, I was pleasantly surprised to see that GraphObject made it to the amd64 platform given that I don't think MS has touched it in a couple of decades. I can't imagine having ported it to 64 bit that they can't just compile it now to the other target. I'm also surprised they're able to ship ...Diagrams.dll without it so I'm curious what the response is.

PS I'm sure you figured out that the formatting stripped the asterisks from my dir statements. I just noticed.

**Reply by CoreyKaylor** [reply]
Looks like their response is that there are no plans to support Arm64...

**Reply by mcurland** [reply]
Do you remember if the ...modeling.sdk... assemblies were there in a default installation or did you have to turn on extension development and the Modeling SDK check to get them? Was this checkbox even available?

I'm curious what _is_ supported by the Modeling SDK checkbox and the installed assemblies (if it is there, which I don't know and don't want to ask the question myself without knowing).

Class designer depends on the modeling files AFAIK, so was probably the driver to port GraphObject to amd64.

**Reply by CoreyKaylor** [reply]
Just to re-clarify. `GraphObject` did not make it. I'm not sure what was there before the installation. I suspect they weren't there and were added after the checkbox. Class designer is also not supported on Arm64.
